[RISCV] Add load/store clustering in post machine schedule

BoyaoWang430 opened the pull request:
#73789 added load clustering and #73796 tried to add store clustering.
If post machine schedule is used, previous cluster of load/store which formed in machine schedule may break. In order to solve this, add load/sotre clustering to post machine schedule.

asb commented:
> #73789 added load clustering and #73796 tried to add store clustering. If post machine schedule is used, previous cluster of load/store which formed in machine schedule may break. In order to solve this, add load/sotre clustering to post machine schedule.

Hi, I'm struggling to parse this a bit. I think you're saying that if load and store clustering aren't added to the post machine schedule in addition to being in `createMachineScheduler`, you see issues. Is that correct? It looks like PPC (which does store clustering only) also adds it in both functions, though AArch64 doesn't.

If you could update this PR to show any test differences that would really help with review. Thanks!

asb commented:
FWIW, if I add a version of this on top of #73796 I get assertions in several llvm/test/CodeGen/RISCV tests

```
+  ScheduleDAGInstrs *
+  createPostMachineScheduler(MachineSchedContext *C) const override {
+    ScheduleDAGMILive *DAG = nullptr;
+    if (EnableMISchedLoadStoreClustering) {
+      DAG = createGenericSchedLive(C);
+      DAG->addMutation(createLoadClusterDAGMutation(
+          DAG->TII, DAG->TRI, /*ReorderWhileClustering=*/true));
+      DAG->addMutation(createStoreClusterDAGMutation(
+          DAG->TII, DAG->TRI, /*ReorderWhileClustering=*/true));
+    }
+    return DAG;
+  }
+
```

Seemingly for some of the tests with SiFive `-mcpu`.

```
0.	Program arguments: /home/asb/llvm-project/build/default/bin/llc -mtriple=riscv64 -mcpu=sifive-u74 -verify-machineinstrs
1.	Running pass 'Function Pass Manager' on module '<stdin>'.
2.	Running pass 'PostRA Machine Instruction Scheduler' on function '@and_select_all_ones_i32'
 #0 0x00007fa8d38393bd llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) /home/asb/llvm-project/llvm/lib/Support/Unix/Signals.inc:723:11
 #1 0x00007fa8d38399ab PrintStackTraceSignalHandler(void*) /home/asb/llvm-project/llvm/lib/Support/Unix/Signals.inc:798:1
 #2 0x00007fa8d3837573 llvm::sys::RunSignalHandlers() /home/asb/llvm-project/llvm/lib/Support/Signals.cpp:105:5
 #3 0x00007fa8d383a241 SignalHandler(int) /home/asb/llvm-project/llvm/lib/Support/Unix/Signals.inc:413:1
 #4 0x00007fa8d2e4c1d0 (/usr/lib/libc.so.6+0x3d1d0)
 #5 0x00007fa8d2ea53f4 (/usr/lib/libc.so.6+0x963f4)
 #6 0x00007fa8d2e4c120 raise (/usr/lib/libc.so.6+0x3d120)
 #7 0x00007fa8d2e334c3 abort (/usr/lib/libc.so.6+0x244c3)
 #8 0x00007fa8d30d3af0 std::chrono::_V2::system_clock::now() /usr/src/debug/gcc/gcc/libstdc++-v3/src/c++11/chrono.cc:52:5
 #9 0x00007fa8d74c5926 std::unique_ptr<llvm::RegisterClassInfo::RCInfo [], std::default_delete<llvm::RegisterClassInfo::RCInfo []>>::operator[](unsigned long) const /usr/bin/../lib64/gcc/x86_64-pc-linux-gnu/14.2.1/../../../../include/c++/14.2.1/bits/unique_ptr.h:726:2
#10 0x00007fa8d74c581e llvm::RegisterClassInfo::get(llvm::TargetRegisterClass const*) const /home/asb/llvm-project/llvm/include/llvm/CodeGen/RegisterClassInfo.h:79:25
#11 0x00007fa8d7b3f28d llvm::RegisterClassInfo::getNumAllocatableRegs(llvm::TargetRegisterClass const*) const /home/asb/llvm-project/llvm/include/llvm/CodeGen/RegisterClassInfo.h:95:20
#12 0x00007fa8d7b24d40 llvm::GenericScheduler::initPolicy(llvm::MachineInstrBundleIterator<llvm::MachineInstr, false>, llvm::MachineInstrBundleIterator<llvm::MachineInstr, false>, unsigned int) /home/asb/llvm-project/llvm/lib/CodeGen/MachineScheduler.cpp:3293:50
#13 0x00007fa8d7b18f1d llvm::ScheduleDAGMI::enterRegion(llvm::MachineBasicBlock*, llvm::MachineInstrBundleIterator<llvm::MachineInstr, false>, llvm::MachineInstrBundleIterator<llvm::MachineInstr, false>, unsigned int) /home/asb/llvm-project/llvm/lib/CodeGen/MachineScheduler.cpp:799:1
#14 0x00007fa8d7b1c178 llvm::ScheduleDAGMILive::enterRegion(llvm::MachineBasicBlock*, llvm::MachineInstrBundleIterator<llvm::MachineInstr, false>, llvm::MachineInstrBundleIterator<llvm::MachineInstr, false>, unsigned int) /home/asb/llvm-project/llvm/lib/CodeGen/MachineScheduler.cpp:1242:18
#15 0x00007fa8d7b2a7a2 (anonymous namespace)::MachineSchedulerBase::scheduleRegions(llvm::ScheduleDAGInstrs&, bool) /home/asb/llvm-project/llvm/lib/CodeGen/MachineScheduler.cpp:645:13
#16 0x00007fa8d7b2bcc3 (anonymous namespace)::PostMachineScheduler::runOnMachineFunction(llvm::MachineFunction&) /home/asb/llvm-project/llvm/lib/CodeGen/MachineScheduler.cpp:514:7
#17 0x00007fa8d79d671c llvm::MachineFunctionPass::runOnFunction(llvm::Function&) /home/asb/llvm-project/llvm/lib/CodeGen/MachineFunctionPass.cpp:94:8
#18 0x00007fa8d4327105 llvm::FPPassManager::runOnFunction(llvm::Function&) /home/asb/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:1408:23
#19 0x00007fa8d432c36f llvm::FPPassManager::runOnModule(llvm::Module&) /home/asb/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:1454:16
#20 0x00007fa8d4327aeb (anonymous namespace)::MPPassManager::runOnModule(llvm::Module&) /home/asb/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:1523:23
#21 0x00007fa8d43275d4 llvm::legacy::PassManagerImpl::run(llvm::Module&) /home/asb/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:541:16
#22 0x00007fa8d432c671 llvm::legacy::PassManager::run(llvm::Module&) /home/asb/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:1650:3
#23 0x0000619ce90cee72 compileModule(char**, llvm::LLVMContext&) /home/asb/llvm-project/llvm/tools/llc/llc.cpp:755:9
#24 0x0000619ce90cceb2 main /home/asb/llvm-project/llvm/tools/llc/llc.cpp:412:13
#25 0x00007fa8d2e34e08 (/usr/lib/libc.so.6+0x25e08)
#26 0x00007fa8d2e34ecc __libc_start_main (/usr/lib/libc.so.6+0x25ecc)
#27 0x0000619ce90cc595 _start (/home/asb/llvm-project/build/default/bin/llc+0x48595)
FileCheck error: '<stdin>' is empty.
FileCheck command line:  /home/asb/llvm-project/build/default/bin/FileCheck -check-prefix=SFB64 /home/asb/llvm-project/llvm/test/CodeGen/RISCV/select-binop-identity.ll
```

wangpc-pp commented:
> createGenericSchedLive

I think the reason is you are using pre-ra scheduler here.

wangpc-pp commented:
> > > createGenericSchedLive
> > 
> > 
> > I think the reason is you are using pre-ra scheduler here.
> 
> Thank you - the problem was indeed me being too eager with copy and paste and failing to switch to `createGenericSchedPostRA`. Now I've done that though, I see zero codegen changes within the in-tree unit tests.

Thanks! I will ask @BoyaoWang430 to add some MIR tests. And thanks in advance if you can help to evaluate/review this PR. :-)

asb commented:
> > > > createGenericSchedLive
> > > 
> > > 
> > > I think the reason is you are using pre-ra scheduler here.
> > 
> > 
> > Thank you - the problem was indeed me being too eager with copy and paste and failing to switch to `createGenericSchedPostRA`. Now I've done that though, I see zero codegen changes within the in-tree unit tests.
> 
> Thanks! I will ask @BoyaoWang430 to add some MIR tests. And thanks in advance if you can help to evaluate/review this PR. :-)

Thanks! And more generally, any note you have on how/if it affects codegen on external codebases very welcome. e.g. is this something that kicks in a lot in real-world code but we just don't trigger in our tests, or is it fairly rare it makes a difference (but of course worth addressing for the cases it helps)

BoyaoWang430 commented:
> > > createGenericSchedLive
> > 
> > 
> > I think the reason is you are using pre-ra scheduler here.
> 
> Thank you - the problem was indeed me being too eager with copy and paste and failing to switch to `createGenericSchedPostRA`. Now I've done that though, I see zero codegen changes within the in-tree unit tests.

We also made the same mistake when we first tried to add this by copy-paste. I will add some MIR tests these days to demonstrate the difference that enabling load/store clustering in the post machine scheduler can make.

wangpc-pp commented:
> > > > > createGenericSchedLive
> > > > 
> > > > 
> > > > I think the reason is you are using pre-ra scheduler here.
> > > 
> > > 
> > > Thank you - the problem was indeed me being too eager with copy and paste and failing to switch to `createGenericSchedPostRA`. Now I've done that though, I see zero codegen changes within the in-tree unit tests.
> > 
> > 
> > Thanks! I will ask @BoyaoWang430 to add some MIR tests. And thanks in advance if you can help to evaluate/review this PR. :-)
> 
> Thanks! And more generally, any note you have on how/if it affects codegen on external codebases very welcome. e.g. is this something that kicks in a lot in real-world code but we just don't trigger in our tests, or is it fairly rare it makes a difference (but of course worth addressing for the cases it helps)

The reason why there is no CodeGen change in in-tree tests is because postra scheduler is not enabled by default. To enable postra scheduler, we should use a CPU with `FeaturePostRAScheduler` or add `+use-postra-scheduler` to `-mattr`.

The problem is very common, maybe I didn't make it clear in sync-up meeting because of my bad speaking :-(. 
* We have added load/store clustering mutations to pre-ra scheduler, so when doing pre-ra scheduling, we can add dependencies between load/store instructions and make them together.
* But, if we enable post-ra scheduling, and load/store clustering mutations are not added to post-ra scheduler, then there is no strong dependency between these load/store instructions and the scheduler may separate them.
For example:
```asm
// Before pre-ra scheduling
ld vreg0, 0(rs)
addi vreg1, vreg0, 1
ld vreg2, 8(rs)
addi vreg3, vreg2, 1

// After pre-ra scheduling
ld vreg0, 0(rs)
ld vreg2, 8(rs)
addi vreg1, vreg0, 1
addi vreg3, vreg2, 1

// After post-ra scheduling (possile current result)
ld a1, 0(a0)
addi a3, a1, 1
ld a2, 8(a0)
addi a4, a2, 1

// After post-ra scheduling (what we want)
ld a1, 0(a0)
ld a2, 8(a0)
addi a3, a1, 1
addi a4, a2, 1
```

I think this problem is common accoss targets and as what you have said PPC have already done it.

topperc commented:
Why is this what we want?

```
ld a1, 0(a0)
ld a2, 8(a0)
addi a3, a1, 1
addi a4, a2, 1
```

What is problem if there is an add between them? Is there some fusion going on in the hardware? I understand why you want them loads ordered near each other if they access the same cache line, but I don't understand why arithmetic between them is bad?

wangpc-pp commented:
> Why is this what we want?
> 
> ```
> ld a1, 0(a0)
> ld a2, 8(a0)
> addi a3, a1, 1
> addi a4, a2, 1
> ```
> 
> What is problem if there is an add between them? Is there some fusion going on in the hardware? I understand why you want them loads ordered near each other if they access the same cache line, but I don't understand why arithmetic between them is bad?

Oh, sorry, my example may not be exact. Several reasons:
1. If the result of enabling pre-ra scheduling is what we are expecting, then I think we should get the same result after post-ra scheduling, right?
2. Some μ-arch things require this:
  * As what you have pointed out, fusion is one thing.
  * Another thing is about store clustering. Store buffer may be impacted as we interleave arithmetic/load(which is not shown in my example) between stores.

Anyway, we saw the benefit and if it is not common for all μ-arch, we can make it a feature.

BoyaoWang430 committed:
[RISCV] Add load/store clustering in post machine schedule

If post machine schedule is used, previous cluster of load/store which formed in machine schedule may break. In order to solve this, add load/sotre clustering to post machine schedule.

wangpc-pp commented on llvm/lib/Target/RISCV/RISCVTargetMachine.cpp:
```diff
@@ -363,6 +363,19 @@ class RISCVPassConfig : public TargetPassConfig {
+  ScheduleDAGInstrs *
+  createPostMachineScheduler(MachineSchedContext *C) const override {
+    ScheduleDAGMI *DAG = nullptr;
+    if (EnableMISchedLoadStoreClustering) {
```
Can we add a new debug option to control if we should enable postra clustering?
```
static cl::opt<bool> EnablePostMISchedLoadStoreClustering(
    "riscv-postmisched-load-store-clustering", cl::Hidden,
    cl::desc("Enable PostRA load and store clustering in the machine scheduler"),
    cl::init(true));

if (EnableMISchedLoadStoreClustering && EnablePostMISchedLoadStoreClustering) {
   ...
}
```

wangpc-pp commented on llvm/test/CodeGen/RISCV/misched-mem-clustering.mir:
```diff
@@ -0,0 +1,64 @@
+# RUN: llc -mtriple=riscv64 -x mir -mcpu=sifive-p470 -mattr=+use-postra-scheduler -verify-misched -enable-post-misched=true \
+# RUN:     -debug-only=machine-scheduler \
+# RUN:     -start-before=machine-scheduler -stop-after=postmisched -o - 2>&1 < %s \
+# RUN:   | FileCheck -check-prefix=MEMCLUSTER %s
```
After adding the option above, please add a RUN with postra scheduling enabled but no postra clustering.

topperc reviewed (approved):
LGTM

wangpc-pp commented:
> LLVM Buildbot has detected a new failure on builder `fuchsia-x86_64-linux` running on `fuchsia-debian-64-us-central1-a-1` while building `llvm` at step 4 "annotate".
> 
> Full details are available at: https://lab.llvm.org/buildbot/#/builders/11/builds/7711
> Here is the relevant piece of the build log for the reference

Should be fixed by 37ce18951fded6be1de319b05b968918cb45c00b.
